crypto/ec/ecp_secp384r1.c

```diff
@@ -0,0 +1,19594 @@
+    fiat_secp384r1_mulx_u64(&x365, &x366, x353, UINT32_C(0xffffffff));
+    fiat_secp384r1_addcarryx_u64(&x367, &x368, 0x0, x366, x363);
+    fiat_secp384r1_addcarryx_u64(&x369, &x370, x368, x364, x361);
+    fiat_secp384r1_addcarryx_u64(&x371, &x372, x370, x362, x359);
+    fiat_secp384r1_addcarryx_u64(&x373, &x374, x372, x360, x357);
+    fiat_secp384r1_addcarryx_u64(&x375, &x376, x374, x358, x355);
+    fiat_secp384r1_addcarryx_u64(&x377, &x378, 0x0, x341, x365);
+    fiat_secp384r1_addcarryx_u64(&x379, &x380, x378, x343, x367);
+    fiat_secp384r1_addcarryx_u64(&x381, &x382, x380, x345, x369);
+    fiat_secp384r1_addcarryx_u64(&x383, &x384, x382, x347, x371);
+    fiat_secp384r1_addcarryx_u64(&x385, &x386, x384, x349, x373);
+    fiat_secp384r1_addcarryx_u64(&x387, &x388, x386, x351, x375);
+    fiat_secp384r1_addcarryx_u64(&x389, &x390, x388, ((uint64_t)x352 + x324),
+                                 (x376 + x356));
+    fiat_secp384r1_subborrowx_u64(&x391, &x392, 0x0, x379,
+                                  UINT32_C(0xffffffff));
+    fiat_secp384r1_subborrowx_u64(&x393, &x394, x392, x381,
+                                  UINT64_C(0xffffffff00000000));
+    fiat_secp384r1_subborrowx_u64(&x395, &x396, x394, x383,
+                                  UINT64_C(0xfffffffffffffffe));
+    fiat_secp384r1_subborrowx_u64(&x397, &x398, x396, x385,
+                                  UINT64_C(0xffffffffffffffff));
+    fiat_secp384r1_subborrowx_u64(&x399, &x400, x398, x387,
+                                  UINT64_C(0xffffffffffffffff));
+    fiat_secp384r1_subborrowx_u64(&x401, &x402, x400, x389,
+                                  UINT64_C(0xffffffffffffffff));
+    fiat_secp384r1_subborrowx_u64(&x403, &x404, x402, x390, 0x0);
+    fiat_secp384r1_cmovznz_u64(&x405, x404, x391, x379);
+    fiat_secp384r1_cmovznz_u64(&x406, x404, x393, x381);
+    fiat_secp384r1_cmovznz_u64(&x407, x404, x395, x383);
+    fiat_secp384r1_cmovznz_u64(&x408, x404, x397, x385);
+    fiat_secp384r1_cmovznz_u64(&x409, x404, x399, x387);
+    fiat_secp384r1_cmovznz_u64(&x410, x404, x401, x389);
+    out1[0] = x405;
+    out1[1] = x406;
+    out1[2] = x407;
+    out1[3] = x408;
+    out1[4] = x409;
+    out1[5] = x410;
+}
+
+/*
+ * The function fiat_secp384r1_nonzero outputs a single non-zero word if the input is non-zero and zero otherwise.
+ * Preconditions:
+ *   0 ≤ eval arg1 < m
+ * Postconditions:
+ *   out1 = 0 ↔ eval (from_montgomery arg1) mod m = 0
+ *
+ * Input Bounds:
+ *   arg1: [[0x0 ~> 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff], [0x0 ~> 0xffffffffffffffff]]
+ * Output Bounds:
+ *   out1: [0x0 ~> 0xffffffffffffffff]
+ */
+static void fiat_secp384r1_nonzero(uint64_t *out1, const uint64_t arg1[6]) {
+    uint64_t x1;
+    x1 = ((arg1[0]) |
+          ((arg1[1]) |
+           ((arg1[2]) |
+            ((arg1[3]) | ((arg1[4]) | ((arg1[5]) | (uint64_t)0x0))))));
```

Comment:
This looks a little weird. I wonder why there is an `| 0`, that doesn't seem to be doing much.

I'm also not sure why all the parentheses are there. Is the order important?

Reply:
Regarding the `| 0`, the template definition is much simpler if we include it; see https://github.com/mit-plv/fiat-crypto/blob/37daa2d8bb34e3e0a30c7396952226407e5e19a3/src/Arithmetic/WordByWordMontgomery.v#L122-L123
And we just didn't have a rule for eliminating this previously.
I've added such a rule in https://github.com/mit-plv/fiat-crypto/commit/37daa2d8bb34e3e0a30c7396952226407e5e19a3, so if you regenerate the file, it should drop the `| 0`.

As far as the parentheses, we always emit parentheses around operations because it's simpler that way; if we didn't do this, we'd have to add in (unverified) pretty-printing logic for associativity and levels of operators (and there's not a universal convention across languages, for example, for which of `&` and `|` binds more tightly).